src/net/atlas/db.rs

```diff
@@ -231,30 +257,76 @@ impl AtlasDB {
         Ok(res)
     }
 
-    pub fn insert_new_attachment(&mut self, attachment: &Attachment) -> Result<(), db_error> {
+    pub fn insert_uninstanciated_attachment(
+        &mut self,
+        attachment: &Attachment,
+    ) -> Result<(), db_error> {
+        // Insert the new attachment
+        let uninstantiated_attachments = self.count_uninstantiated_attachments()?;
+        if uninstantiated_attachments >= self.atlas_config.max_uninstanciated_attachments {
+            let to_delete =
+                1 + uninstantiated_attachments - self.atlas_config.max_uninstanciated_attachments;
+            self.evict_k_oldest_uninstantiated_attachments(to_delete)?;
+        }
+
         let tx = self.tx_begin()?;
+        let now = util::get_epoch_time_secs() as i64;
         let res = tx.execute(
-            "INSERT OR REPLACE INTO attachments (hash, content, was_instantiated) VALUES (?, ?, 0)",
+            "INSERT OR REPLACE INTO attachments (hash, content, was_instantiated, created_at) VALUES (?, ?, 0, ?)",
             &[
                 &attachment.hash() as &dyn ToSql,
                 &attachment.content as &dyn ToSql,
+                &now as &dyn ToSql,
             ],
         );
         res.map_err(db_error::SqliteError)?;
         tx.commit().map_err(db_error::SqliteError)?;
         Ok(())
     }
 
+    pub fn evict_k_oldest_uninstantiated_attachments(&mut self, k: u32) -> Result<(), db_error> {
+        let tx = self.tx_begin()?;
+        let res = tx.execute(
+            "DELETE FROM attachments WHERE hash IN (SELECT hash FROM attachments WHERE was_instantiated = 0 ORDER BY created_at ASC LIMIT ?)",
+            &[&k as &dyn ToSql],
+        );
+        res.map_err(db_error::SqliteError)?;
+        tx.commit().map_err(db_error::SqliteError)?;
+        Ok(())
+    }
+
+    pub fn evict_expired_uninstantiated_attachments(&mut self) -> Result<(), db_error> {
+        let now = util::get_epoch_time_secs() as i64;
+        let cut_off = now - self.atlas_config.uninstanciated_attachments_expire_after as i64;
+        let tx = self.tx_begin()?;
+        let res = tx.execute(
+            "DELETE FROM attachments WHERE was_instantiated = 0 AND created_at < ?",
+            &[&cut_off as &dyn ToSql],
+        );
+        res.map_err(db_error::SqliteError)?;
+        tx.commit().map_err(db_error::SqliteError)?;
+        Ok(())
+    }
+
+    pub fn count_uninstantiated_attachments(&self) -> Result<u32, db_error> {
+        let qry = "SELECT COUNT(*) FROM attachments
```

Comment:
`COUNT(*)` is a _O(n)_ operation in Sqlite.  Is there a way to avoid this call, or minimize its usage?  Over time, it will get more and more expensive.

If you're just looking to see if there are more than _k_ uninstantiated attachments, one way to do it is to create an index on the `was_instantiated` column and then do a `SELECT was_instantiated FROM attachments WHERE was_instantiated = 0 LIMIT ${k}`.  Then, it's a _O(k log n)_ operation, since you're doing an indexed B+-tree scan.

Reply:
> Over time, it will get more and more expensive. 

But it is bound by the `WHERE` condition so `max_uninstanciated_attachments`, no?

Reply:
Also, with the suggested SQL query, won't we end up returning / validating / instantiating N `i64`? Isn't better to just let sqlite doing a `COUNT`?

Reply:
> But it is bound by the WHERE condition so max_uninstantiated_attachments, no?

Only if you create an index that maps `was_instantiated` to `attachments.hash`.  Otherwise it's a full table scan.  Also, `COUNT(ROWID)` is much faster than `COUNT(*)`.  You'll want to verify this with `EXPLAIN QUERY PLAN`.

Reply:
Thank you @jcnelson.
Since it's more usable, I went with 
```sql
SELECT COUNT(rowid) FROM attachments WHERE was_instantiated = 0;
```
that has the same query plan than

```sql
SELECT was_instantiated FROM attachments WHERE was_instantiated = 0 LIMIT N
```

as long as you add an index on `was_instantiated`

```bash
sqlite> EXPLAIN QUERY PLAN SELECT COUNT(*) FROM attachments WHERE was_instantiated = 0;
QUERY PLAN
--SEARCH TABLE attachments USING COVERING INDEX index_was_instanciated (was_instantiated=?)

sqlite> EXPLAIN QUERY PLAN SELECT was_instantiated FROM attachments WHERE was_instantiated = 0 LIMIT 2;
QUERY PLAN
--SEARCH TABLE attachments USING COVERING INDEX index_was_instanciated (was_instantiated=?)
```

Reply:
Sounds good :+1: Thanks for checking with the query planner!  As long as we're using a `COVERING INDEX` to do the scan, then it's fine by me.